w1r3/python/w1r3.py

```diff
@@ -0,0 +1,377 @@
+#!/usr/bin/env python
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""The Python version of the W1R3 benchmark."""
+
+import argparse
+import random
+import resource
+import sys
+import time
+import uuid
+
+import google.cloud.storage as storage
+import google.protobuf
+import grpc
+import opentelemetry.metrics as metrics
+import opentelemetry.sdk.metrics as sdkmetrics
+import opentelemetry.sdk.metrics.view as view
+import opentelemetry.sdk.trace as sdktrace
+import opentelemetry.trace as trace
+import requests
+from opentelemetry.exporter.cloud_monitoring import CloudMonitoringMetricsExporter
+from opentelemetry.exporter.cloud_trace import CloudTraceSpanExporter
+from opentelemetry.instrumentation.requests import RequestsInstrumentor
+from opentelemetry.resourcedetector.gcp_resource_detector import (
+    GoogleCloudResourceDetector,
+)
+from opentelemetry.sdk.metrics.export import PeriodicExportingMetricReader
+from opentelemetry.sdk.resources import Resource
+from opentelemetry.sdk.trace.export import BatchSpanProcessor
+
+# You can optionally pass a custom TracerProvider to instrument().
+response = requests.get(url="https://www.example.org/")
+
+
+def main():
+    args = parse_args()
+    instance = str(uuid.uuid4())
+    print("## Starting continuous GCS Python SDK benchmark")
+    print(f"# object-sizes: {args.object_size}")
+    print(f"# transports: {args.transport}")
+    print(f"# project-id: {args.project_id}")
+    print(f"# bucket: {args.bucket}")
+    print(f"# deployment: {args.deployment}")
+    print(f"# instance: {instance}")
+    print(f"# Benchmark Version: {_SSB_W1R3_VERSION}")
+    print(f"# Python Version: {sys.version}")
+    print(f"# Python SDK Version: {storage.__version__}")
+    print(f"# gRPC Version: {grpc.__version__}")
+    print(f"# Protobuf Version: {google.protobuf.__version__}")
+    print(f"# HTTP Client Version: {requests.__version__}")
+
+    tracer = configure_trace_exporter(args)
+    (meter_provider, meter) = configure_metrics_exporter(instance, args)
+
+    # Instrument Python's `requests` library to generate spans.
+    RequestsInstrumentor().instrument()
+
+    clients = make_clients(args)
+    prng = random.Random()
+    prng.seed()
+    random_data = bytes([prng.randint(0, 255) for _ in range(0, max(args.object_size))])
+
+    worker(clients, instance, prng, random_data, tracer, meter, args)
```

Comment:
It seems like we’re running a worker in a single process. Will this be ran concurrently?

Reply:
Yes, we end up running N processes on the same VM.  We use a single thread per-process to make CPU (and eventually memory) measurements easier.